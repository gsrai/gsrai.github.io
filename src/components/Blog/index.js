import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

class Blog extends React.Component {
  constructor(props) {
    super(props)
    this.state = { posts: null, error: false }
  }

  async componentWillMount() {
    try {
      const posts = await this.getPosts()
      this.setState({ posts })
    } catch (e) {
      console.error(e)
      this.setState({ error: true })
    }
  }

  getPosts = async () => {
    const endpoint = 'assets/posts/manifest.json'
    const res = await fetch(endpoint)
    const json = await res.json()
    return json.posts
  }

  render() {
    const { posts, error } = this.state
    return (
      <div>
        <div className='row flex-center'>
          { posts
            ? posts.map((post, i) => (<BlogPostCard key={i} {...post} />))
            : 'loading...' }
          { error && 'No Posts'}
        </div>
      </div>
    )
  }
}

export const BlogPostCard = ({ title, published, description, url }) => {
  /* assets/posts/ is 13 characters, -3 to remove the .md extension */
  const filename = url.slice(13).slice(0, -3)
  const linkUrl = `/blog/posts/${filename}`
  return (
    <div className='card' style={{ width: '20rem', marginBottom: '2rem' }}>
      <div className='card-body'>
        <h4 className='card-title'>{title}</h4>
        <h5 className='card-subtitle'>{published}</h5>
        <p className='card-text'>{description}</p>
        <Link className='card-link' to={linkUrl}>read more</Link>
      </div>
    </div>
  )
}

BlogPostCard.propTypes = {
  title: PropTypes.string.isRequired,
  published: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
}

export default Blog
