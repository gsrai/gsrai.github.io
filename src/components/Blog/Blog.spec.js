import { shallow } from 'enzyme'
import React from 'react'
// import { MemoryRouter, Link } from 'react-router-dom'
import Blog, { BlogPostCard } from './index'

describe('Component Rendering', () => {
  test('Blog component should render correctly', () => {
    const wrapper = shallow(<Blog />)
    expect(wrapper).toMatchSnapshot()
  })

  test('BlogPostCard component should render correctly', () => {
    const props = {
      title: 'why we moved from language x to language y',
      published: '22/11/2018',
      description: 'We stopped using language x, now we use language y, you should do the same.',
      url: 'some/url'
    }
    const wrapper = shallow(<BlogPostCard {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})

// describe('Component Behavior', () => {
//   test('renders Links correctly', () => {
//     const wrapper = mount(
//       <MemoryRouter>
//         <Header />
//       </MemoryRouter>
//     )

//     const linksTo = wrapper.find(Link).map(l => l.props().to)
//     expect(linksTo).toEqual(['/', '/about', '/blog'])
//   })
//   test('Active link should have correct className', () => {
//     const classname = '.btn-success'
//     const wrapper = mount(
//       <MemoryRouter>
//         <Header />
//       </MemoryRouter>
//     )

//     expect(wrapper.find(classname).exists()).toBe(true)
//   })
// })
