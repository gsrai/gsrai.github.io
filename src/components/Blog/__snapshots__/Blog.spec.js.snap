// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Rendering Blog component should render correctly 1`] = `
<div>
  <div
    className="row flex-center"
  >
    loading...
  </div>
</div>
`;

exports[`Component Rendering BlogPostCard component should render correctly 1`] = `
<div
  className="card"
  style={
    Object {
      "marginBottom": "2rem",
      "width": "20rem",
    }
  }
>
  <div
    className="card-body"
  >
    <h4
      className="card-title"
    >
      why we moved from language x to language y
    </h4>
    <h5
      className="card-subtitle"
    >
      22/11/2018
    </h5>
    <p
      className="card-text"
    >
      We stopped using language x, now we use language y, you should do the same.
    </p>
    <Link
      className="card-link"
      replace={false}
      to="/blog/posts/"
    >
      read more
    </Link>
  </div>
</div>
`;
